### ISSUE TITLE
Context menu Delete X blocks preview does not match behavior

### ISSUE THREAD

**Issue by BeksOmega** [google/blockly#4832]
**Describe the bug**

When you open up the context menu and select the "Delete X blocks" option, X blocks may not actually be deleted, because the stack is healed.
![delete-blocks](https://user-images.githubusercontent.com/25440652/118566439-b947d700-b728-11eb-8e92-7c36bb5046cd.gif)

**To Reproduce**

1. Load the following XML into the playground:
```
<xml xmlns="https://developers.google.com/blockly/xml">
  <variables>
    <variable>test</variable>
  </variables>
  <block type="variables_set">
    <field name="VAR">test</field>
    <value name="VALUE">
      <block type="text_join">
        <mutation items="2"></mutation>
        <value name="ADD0">
          <block type="text">
            <field name="TEXT"></field>
          </block>
        </value>
      </block>
    </value>
  </block>
</xml>
```
2) Right click on the middle block and select "Delete 2 Blocks"
3) Observe how only one block is deleted due to healing the stack.

**Expected behavior**

Either both blocks should be deleted, or the context menu should show the correct preview. I recommend the former, because you should be able to achieve that easily by passing healStack as false to dispose.

**Additional context**

Reported here: https://groups.google.com/g/blockly/c/rRB1IJy_iWo

**Comment by jschanker** [comment]
I agree with Beka's suggestion about deleting the block.  It might also make sense to have this function include two parameters, one for whether to attempt to move up the stack of next blocks and another for whether to attempt to move attached input values across.  The variable name heal*Stack* along with the documentation only mentioning healing any gap by connecting the next statement with the previous one seems to suggest that the original intention might have been to only close the vertical gap and closing the horizontal one was added later?  Unfortunately, as Beka pointed out to me, the subclass `BlockSvg` is already using a Boolean as the second parameter variable to control whether there should be an optional disposal animation.  But a workaround would be to check if the third parameter variable isn't `undefined` and if so to swap the last two variables.  Thoughts?

**Comment by alschmiedt** [comment]
Thanks for reporting!

Talking to the team and looking through old PRs this is almost certainly a bug. It looks like healing a stack was added later as a feature for insertion markers and so this is almost definitely an unfortunate side effect. 

I think we will probably need to discuss the desired behavior a bit more. It seems like healing a stack is an intended feature rather than a bug so I would want to try to understand more about why it was added before removing. If we decide that this feature is no longer needed then I think Beka's recommendation of passing healStack as false in dispose will work nicely, if not then we will probably want to take a look at doing something like Jason recommended. I've added this to our Monday meeting so we should have more information next week.

**Comment by alschmiedt** [comment]
Ok, so we decided that the blocks that are connected across should not heal the stack, but blocks that are connected in a stack should be healed. The reasoning behind this was that blocks that are built across generally have more type checks so there are more cases where trying to heal the stack will fail and also we generally see the blocks built across as a single entity so it makes sense for them to be deleted at once. If a developer wants to change this functionality it is possible to do so by creating their own context menu item and keyboard shortcut.

I think what this means for this bug is that we need some separation between healing a row and healing a stack. Which is what Jason mentioned in his comment. I don't love the idea having to swap the variables, but I have yet to look into a better solution so don't have any other recommendations at the moment : )

**Comment by jschanker** [comment]
Thanks for the update, Abby!  Although I proposed it, this workaround was also not very appealing to me.  However, I thought it'd be good to give developers the choice of whether to try to heal horizontally and/or vertically since the row/stack unplugging logic is already cleanly separated in private methods.  But thinking about it more, I realized that we can already do this without any monkey patching since blocks with output connections will never result in vertical stack healing while ones without output connections will never result in row healing.  So since deleting a block can only result in at most one type of healing, we can pass `true/false` to `dispose` based on the block's connections accordingly to get the combination of healing that we want.  So for the [callback](https://github.com/google/blockly/blob/master/core/contextmenu_items.js#L480) of `Blockly.ContextMenuItems.registerDelete`, it seems to me that we could use the following to permit only vertical stack healing:

```
if(scope.block.outputConnection) {
  scope.block.dispose(false, true);
} else {
  scope.block.dispose(true, true);
}
```

As an aside, I had probably gotten the idea of switching parameter variables because I had previously seen this done in the deprecation of `domToWorkspace` (https://github.com/google/blockly/commit/4e42a1b78ee7bce8f6c4ae8a6600bfc6dbcc3209#comments).
